### ISSUE TITLE
Memory leak when using `StreamingDataset`'s `__iter__` method.

### ISSUE THREAD

**Issue by wadimiusz** [mosaicml/streaming#264]
Hello there,
I have a rather large dataset (around 1.5T) in my GCS that I'm trying to load with `StreamingDataset`, and when I'm trying to access it's `__iter__` method, the Python session gets killed with what appears to be a memory leak. I note that random access to any sample works - it's only `__iter__` that causes a problem.

For me, the following code works:

```
from streaming import StreamingDataset

os.environ["GCS_KEY"] = # value
os.environ["GCS_SECRET"] = # value

dataset = StreamingDataset(remote="gs://some/dataset/path/", shuffle=True)
obj = dataset[2500000]  # random access works fine!
```

However, any of the following breaks:

```
import torch.utils.data
dl = torch.utils.data.DataLoader(dataset)
for batch in dl:
   break

# trying to iterate over the dataloader breaks (the dataloader calls `__iter__`)
```

Or:
```
import torch.utils.data
dl = torch.utils.data.DataLoader(dataset)
obj = next(dl.__iter__())
# Calling __iter__ directly also breaks
```

When the Python process gets killed, I see the following message in my console:

```
/usr/local/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 4 leaked shared_memory objects to clean up at shutdown
```

I'm not sure how to make this example reproducible - it probably works for your datasets, and it's only my system that's weirdly messed up, but I don't know what additional information to include. 
Some data points, though:
1) the bug persists whether or not I use shuffle, or what shuffling algo I use,
2) the bug persists regardless of whether or not I specify a local caching directory,
3) the bug disappears when I try to reproduce it with some tiny dataset.

I wrote the dataset with an MDSWriter.

**Comment by karan6181** [comment]
@wadimiusz Thanks for trying out Streaming Dataset and thanks for narrowing down the issue. I have few questions:
- What is your machine configuration on which you are trying this out ?
- Do you see any traceback when you use `iter` ?
- Can you please let us know the parameter value that are getting passed to `get_partitions_orig()` method? Such as `num_samples` which is I think `6439707754`, then `num_canonical_nodes`, `num_physical_nodes`, `ranks_per_node`, `workers_per_rank`, `batch_size`, and `drop_first`.

Thanks!

**Comment by wadimiusz** [comment]
@karan6181 thanks for your reply!
1. I tried this out on several machines and the bug reproduced. For debugging purposes, I used a machine with 120 GB RAM and 30 CPUs (crazy, right? and still an OOM!). The disk size is 500 GB. It runs on Ubuntu 22.04.2. 
The bug was reproduced on smaller machines with Debian and MacOS. I don't know if you need their specifications, I thought I'd give you the info on the biggest one.

2. No, it just writes "killed" and the Python session ends abruptly.

3. Here are the values:

num_samples = 6439707754
num_canonical_nodes = 1
num_physical_nodes = 1
ranks_per_node = 1
workers_per_rank = 1
batch_size = None
drop_first = 0

PS This is one node, but I'm planning to use it in a multi-node way when I've figured out how to make it work...

**Comment by karan6181** [comment]
Hi @wadimiusz , The number of samples are huge, around 6.5 Billon. Can you try running it again using `streaming-dataset` version `0.5.2` ? I directly ran `python streaming/base/partition/orig.py` with the above parameters and I am able to run it successfully. For me, it uses around 100GB of RAM.

**Comment by karan6181** [comment]
@wadimiusz, I am closing this issue for now due to inactivity. Please don't hesitate to re-open if you still see the issue with the latest streaming version.

**Comment by huxuan** [comment]
I encountered a similar issue with the latest version 0.7.6. After a brief investigation with [tracemalloc](https://docs.python.org/3/library/tracemalloc.html), it reports that there is one line that consumed the most memory with no further trackback.

![image](https://github.com/user-attachments/assets/83128966-de8f-461d-980c-23d97720ca95)
